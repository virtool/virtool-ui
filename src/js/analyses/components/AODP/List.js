import React from "react";
import AnalysisViewerList from "../Viewer/List";
import AODPItem from "./Item";

export const AODPList = () => (
    <AnalysisViewerList itemSize={38} width={300}>
        {AODPItem}
    </AnalysisViewerList>
);
