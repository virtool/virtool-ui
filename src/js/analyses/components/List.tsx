import { map, sortBy } from "lodash-es";
import React from "react";
import { connect } from "react-redux";
import { ContainerNarrow, NoneFoundBox } from "../../base";
import CreateAnalysis from "./Create/CreateAnalysis";
import AnalysisHMMAlert from "./HMMAlert";
import AnalysisItem from "./Item";
import AnalysesToolbar from "./Toolbar";

interface AnalysesListProps {
    analyses: any[];
}

function AnalysesList({ analyses }: AnalysesListProps) {
    return (
        <ContainerNarrow>
            <AnalysisHMMAlert />
            <AnalysesToolbar />

            {analyses.length ? (
                map(sortBy(analyses, "created_at").reverse(), document => (
                    <AnalysisItem key={document.id} {...document} />
                ))
            ) : (
                <NoneFoundBox noun="analyses" />
            )}

            <CreateAnalysis />
        </ContainerNarrow>
    );
}

function mapStateToProps(state) {
    return {
        analyses: state.analyses.documents,
    };
}

export default connect(mapStateToProps)(AnalysesList);
