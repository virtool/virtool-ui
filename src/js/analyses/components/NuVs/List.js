import React from "react";
import AnalysisViewerList from "../Viewer/List";
import NuVsItem from "./Item";

export const NuVsList = () => (
    <AnalysisViewerList itemSize={58} width={230}>
        {NuVsItem}
    </AnalysisViewerList>
);
