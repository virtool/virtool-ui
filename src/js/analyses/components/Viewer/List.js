import { findIndex } from "lodash-es";
import React from "react";
import { connect } from "react-redux";
import { FixedSizeList } from "react-window";
import styled from "styled-components";
import { getBorder, getFontSize } from "../../../app/theme";
import { Key } from "../../../base";
import { setActiveHitId } from "../../actions";
import { getActiveHit, getMatches, getResults } from "../../selectors";
import { useKeyNavigation } from "./hooks";

const AnalysisViewerListHeader = styled.div`
    background-color: ${props => props.theme.color.greyLightest};
    border: 1px solid ${props => props.theme.color.greyLight};
    border-bottom: none;
    border-top-left-radius: ${props => props.theme.borderRadius.sm};
    border-top-right-radius: ${props => props.theme.borderRadius.sm};
    box-shadow: 0 5px 5px -3px ${props => props.theme.color.greyLight};
    padding: 7px 15px;
    z-index: 20;
`;

const AnalysisViewerListFooter = styled.div`
    font-size: ${getFontSize("sm")};
    padding: 15px;
    text-align: center;
`;

const AnalysisViewerListWindow = styled(FixedSizeList)`
    border: ${getBorder};
    border-bottom-left-radius: ${props => props.theme.borderRadius.sm};
    border-bottom-right-radius: ${props => props.theme.borderRadius.sm};
    z-index: 10;
`;

const StyledAnalysisViewerList = styled.div`
    position: relative;
    width: ${props => props.width}px;
`;

export const AnalysisViewerList = ({
    activeId,
    children,
    itemSize,
    nextId,
    nextIndex,
    previousId,
    previousIndex,
    shown,
    total,
    width,
    onSetActiveId
}) => {
    const ref = useKeyNavigation(activeId, nextId, nextIndex, previousId, previousIndex, true, onSetActiveId);

    return (
        <StyledAnalysisViewerList width={width}>
            <AnalysisViewerListHeader>
                Showing {shown} of {total}
            </AnalysisViewerListHeader>
            <AnalysisViewerListWindow ref={ref} height={500} width={width} itemCount={shown} itemSize={itemSize}>
                {children}
            </AnalysisViewerListWindow>
            <AnalysisViewerListFooter>
                Use <Key>w</Key> and <Key>s</Key> to move
            </AnalysisViewerListFooter>
        </StyledAnalysisViewerList>
    );
};

export const mapStateToProps = state => {
    const matches = getMatches(state);
    const active = getActiveHit(state);

    let activeId;
    let nextId;
    let nextIndex;
    let previousId;
    let previousIndex;

    if (active) {
        const activeId = active.id;
        const windowIndex = findIndex(matches, { id: activeId });

        if (windowIndex > 0) {
            previousIndex = windowIndex - 1;
            previousId = matches[previousIndex].id;
        }

        if (windowIndex < matches.length - 1) {
            nextIndex = windowIndex + 1;
            nextId = matches[nextIndex].id;
        }
    }

    return {
        shown: matches.length,
        total: getResults(state).length,
        activeId,
        nextId,
        nextIndex,
        previousId,
        previousIndex
    };
};

export const mapDispatchToProps = dispatch => ({
    onSetActiveId: index => {
        dispatch(setActiveHitId(index));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(AnalysisViewerList);
