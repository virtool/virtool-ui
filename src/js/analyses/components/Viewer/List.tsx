import { findIndex } from "lodash-es";
import React from "react";
import { connect } from "react-redux";
import { FixedSizeList } from "react-window";
import styled from "styled-components";
import { getBorder, getFontSize } from "../../../app/theme";
import { Key } from "../../../base";
import { setActiveHitId } from "../../actions";
import { getActiveHit, getHits, getMatches } from "../../selectors";
import { useKeyNavigation } from "./hooks";

const AnalysisViewerListHeader = styled.div`
    background-color: ${props => props.theme.color.greyLightest};
    border: 1px solid ${props => props.theme.color.greyLight};
    border-bottom: none;
    border-top-left-radius: ${props => props.theme.borderRadius.sm};
    border-top-right-radius: ${props => props.theme.borderRadius.sm};
    box-shadow: 0 5px 5px -3px ${props => props.theme.color.greyLight};
    padding: 7px 15px;
    z-index: 20;
`;

const AnalysisViewerListFooter = styled.div`
    font-size: ${getFontSize("sm")};
    padding: 15px;
    text-align: center;
`;

const AnalysisViewerListWindow = styled(FixedSizeList)`
    border: ${getBorder};
    border-bottom-left-radius: ${props => props.theme.borderRadius.sm};
    border-bottom-right-radius: ${props => props.theme.borderRadius.sm};
    z-index: 10;
`;

type StyledAnalysisViewerListProps = {
    width: number;
};

const StyledAnalysisViewerList = styled.div<StyledAnalysisViewerListProps>`
    position: relative;
    width: ${props => props.width}px;
`;

export function AnalysisViewerList({
    activeId,
    children,
    itemSize,
    nextId,
    nextIndex,
    previousId,
    previousIndex,
    shown,
    total,
    width,
    onSetActiveId,
}) {
    const ref = useKeyNavigation(activeId, nextId, nextIndex, previousId, previousIndex, true, onSetActiveId);

    return (
        <StyledAnalysisViewerList width={width}>
            <AnalysisViewerListHeader>
                Showing {shown} of {total}
            </AnalysisViewerListHeader>
            <AnalysisViewerListWindow ref={ref} height={500} width={width} itemCount={shown} itemSize={itemSize}>
                {children}
            </AnalysisViewerListWindow>
            <AnalysisViewerListFooter>
                Use <Key>w</Key> and <Key>s</Key> to move
            </AnalysisViewerListFooter>
        </StyledAnalysisViewerList>
    );
}

function mapStateToProps(state) {
    const active = getActiveHit(state);
    const matches = getMatches(state);

    let nextId;
    let nextIndex;
    let previousId;
    let previousIndex;

    if (active) {
        const windowIndex = findIndex(matches, { id: active.id });

        if (windowIndex > 0) {
            previousIndex = windowIndex - 1;
            previousId = matches[previousIndex].id;
        }

        if (windowIndex < matches.length - 1) {
            nextIndex = windowIndex + 1;
            nextId = matches[nextIndex].id;
        }
    }
    return {
        activeId: active?.id,
        shown: matches.length,
        total: getHits(state).length,
        nextId,
        nextIndex,
        previousId,
        previousIndex,
    };
}

function mapDispatchToProps(dispatch) {
    return {
        onSetActiveId: index => {
            dispatch(setActiveHitId(index));
        },
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(AnalysisViewerList);
