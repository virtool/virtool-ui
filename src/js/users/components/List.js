/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports UsersList
 */
import React from "react";
import { connect } from "react-redux";
import { LoadingPlaceholder, NoneFoundBox, ScrollList } from "../../base";
import { findUsers } from "../actions";
import { getTerm } from "../selectors";
import UserItem from "./Item";

export class UsersList extends React.Component {
    componentDidMount() {
        this.props.onLoadNextPage(this.props.term, 1);
    }

    renderRow = index => <UserItem key={index} index={index} />;

    render() {
        if (this.props.documents === null) {
            return <LoadingPlaceholder />;
        }

        if (this.props.documents.length) {
            return (
                <ScrollList
                    documents={this.props.documents}
                    onLoadNextPage={page => this.props.onLoadNextPage(this.props.term, page)}
                    page={this.props.page}
                    pageCount={this.props.page_count}
                    renderRow={this.renderRow}
                />
            );
        }

        return <NoneFoundBox noun="users" />;
    }
}

export const mapStateToProps = state => {
    const { documents, page, page_count } = state.users;

    return {
        documents,
        term: getTerm(state),
        page,
        page_count
    };
};

export const mapDispatchToProps = dispatch => ({
    onLoadNextPage: (term, page) => {
        dispatch(findUsers(term, page));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(UsersList);
