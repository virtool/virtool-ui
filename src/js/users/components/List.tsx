import { reduce } from "lodash-es";
import React from "react";
import { useInfiniteFindUsers } from "../../administration/querys";
import { LoadingPlaceholder, NoneFoundBox } from "../../base";
import { StreamlinedScrollList } from "../../base/ScrollList";
import { User } from "../types";
import { UserItem } from "./Item";

function renderRow(item: User) {
    return <UserItem key={item.id} id={item.id} handle={item.handle} administrator_role={item.administrator_role} />;
}

type UsersListProps = {
    term: string;
};

export function UsersList({ term }: UsersListProps) {
    const { data, isLoading, fetchNextPage, isFetchingNextPage } = useInfiniteFindUsers(25, term);

    if (isLoading) {
        return <LoadingPlaceholder />;
    }

    const items = reduce(data.pages, (acc, page) => [...acc, ...page.items], []);

    if (items.length) {
        return (
            <StreamlinedScrollList
                fetchNextPage={fetchNextPage}
                isFetchingNextPage={isFetchingNextPage}
                isLoading={isLoading}
                items={items}
                renderRow={renderRow}
            />
        );
    }

    return <NoneFoundBox noun="users" />;
}
