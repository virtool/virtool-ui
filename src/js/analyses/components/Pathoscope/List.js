import { findIndex, map } from "lodash-es";
import React from "react";
import { connect } from "react-redux";
import { setActiveHitId } from "../../actions";
import { getActiveHit, getMatches, getResults } from "../../selectors";
import { useKeyNavigation } from "../Viewer/hooks";
import PathoscopeItem from "./Item";

export const PathoscopeList = ({ activeId, nextId, nextIndex, previousId, previousIndex, matches, onSetActiveId }) => {
    const ref = useKeyNavigation(activeId, nextId, nextIndex, previousId, previousIndex, false, onSetActiveId);
    const itemComponents = map(matches, (match, index) => <PathoscopeItem key={match.id} index={index} {...match} />);
    return <div ref={ref}>{itemComponents}</div>;
};

export const mapStateToProps = state => {
    const matches = getMatches(state);
    const active = getActiveHit(state);

    let activeId;
    let nextId;
    let nextIndex;
    let previousId;
    let previousIndex;

    if (active) {
        const activeId = active.id;
        const windowIndex = findIndex(matches, { id: activeId });

        if (windowIndex > 0) {
            previousIndex = windowIndex - 1;
            previousId = matches[previousIndex].id;
        }

        if (windowIndex < matches.length - 1) {
            nextIndex = windowIndex + 1;
            nextId = matches[nextIndex].id;
        }
    }

    return {
        shown: matches.length,
        total: getResults(state).length,
        activeId,
        matches,
        nextId,
        nextIndex,
        previousId,
        previousIndex
    };
};

export const mapDispatchToProps = dispatch => ({
    onSetActiveId: index => {
        dispatch(setActiveHitId(index));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(PathoscopeList);
