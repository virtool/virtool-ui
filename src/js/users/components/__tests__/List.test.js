import React from "react";
import { ThemeProvider } from "styled-components";
import { FIND_USERS } from "../../../app/actionTypes";
import { ScrollList } from "../../../base";
import { UsersList, mapStateToProps, mapDispatchToProps } from "../List";

const ThemeProviderWrapper = ({ children }) => {
    const theme = {
        color: {
            greyLight: "#dddddd"
        }
    };

    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const mountWithTheme = tree =>
    mount(tree, {
        wrappingComponent: ThemeProviderWrapper
    });

describe("<UsersList />", () => {
    let props;

    beforeEach(() => {
        props = {
            documents: [{ id: "fred" }],
            term: "foo",
            onLoadNextPage: jest.fn(),
            page: 2,
            page_count: 5
        };
    });

    it("should render", () => {
        const wrapper = shallow(<UsersList {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render when [documents=null]", () => {
        props.documents = null;
        const wrapper = shallow(<UsersList {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should call onLoadNextPage() on mount", () => {
        props.documents = null;
        mountWithTheme(<UsersList {...props} />);
        expect(props.onLoadNextPage).toHaveBeenCalledWith("foo", 1);
    });

    it("should call onLoadNextPage() when paged", () => {
        const wrapper = shallow(<UsersList {...props} />);
        wrapper.find(ScrollList).prop("onLoadNextPage")(8);
        expect(props.onLoadNextPage).toHaveBeenCalledWith("foo", 8);
    });
});

describe("mapStateToProps()", () => {
    it("should return props", () => {
        const users = {
            documents: [{ id: "bob" }],
            page: 2,
            page_count: 10,
            term: "foo"
        };
        const props = mapStateToProps({
            users
        });

        expect(props).toEqual(users);
    });
});

describe("mapDispatchToProps()", () => {
    it("should return onLoadNextPage() in props", () => {
        const dispatch = jest.fn();
        const props = mapDispatchToProps(dispatch);
        const term = "foo";
        const page = 3;

        props.onLoadNextPage(term, page);

        expect(dispatch).toHaveBeenCalledWith({
            type: FIND_USERS.REQUESTED,
            term,
            page
        });
    });
});
