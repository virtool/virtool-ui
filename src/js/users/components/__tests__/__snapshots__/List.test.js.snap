// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UsersList /> should render 1`] = `
<ScrollList
  documents={
    Array [
      Object {
        "id": "fred",
      },
    ]
  }
  onLoadNextPage={[Function]}
  page={2}
  pageCount={5}
  renderRow={[Function]}
/>
`;

exports[`<UsersList /> should render when [documents=null] 1`] = `<LoadingPlaceholder />`;
